apps/builder/app/routes/auth.ws_.callback.ts

```diff
@@ -0,0 +1,70 @@
+import { type LoaderFunction, redirect } from "@remix-run/server-runtime";
+import { AuthorizationError } from "remix-auth";
+import createDebug from "debug";
+import { isBuilderUrl } from "~/shared/router-utils/origins";
+import { returnToPath } from "~/services/cookie.server";
+import { builderAuthenticator } from "~/services/builder-auth.server";
+import { OAuth2Error } from "remix-auth-oauth2";
+import { builderSessionStorage } from "~/services/builder-session.server";
+
+const debug = createDebug("Auth").extend("Ws");
+
+export const loader: LoaderFunction = async ({ request }) => {
+  try {
+    if (false === isBuilderUrl(request.url)) {
+      debug(`Request url is not the builder URL ${request.url}`);
+
+      return new Response(null, {
+        status: 404,
+        statusText: "Only builder URL is allowed",
+      });
+    }
+
+    const returnTo = await returnToPath(request);
+
+    debug("Start exchanging the code for the access token");
+
+    await builderAuthenticator.authenticate("ws", request, {
+      throwOnError: true,
+      successRedirect: returnTo,
+    });
+  } catch (error) {
+    // all redirects are basically errors and in that case we don't want to catch it
+    if (error instanceof Response) {
+      return error;
+    }
+
+    if (error instanceof AuthorizationError) {
+      debug("Authorization error", error.message, error.cause?.message);
+
+      const sessionError = {
+        message: error.message,
+        description: "",
+      };
+
+      if (error.cause instanceof OAuth2Error) {
+        debug(
+          "OAuth2Error error",
+          error.cause?.message,
+          error.cause?.description
+        );
+        sessionError.description = error.cause.description ?? "";
+      }
+
+      const session = await builderSessionStorage.getSession(
+        request.headers.get("Cookie")
+      );
+      session.flash(builderAuthenticator.sessionErrorKey, sessionError);
+
+      throw redirect("/error", {
+        headers: {
+          "Set-Cookie": await builderSessionStorage.commitSession(session),
+        },
+      });
+    }
+
+    debug("error", error);
+    console.error("error", error);
```

Comment:
why do you need both?

Reply:
console shown independent of debug, debug is colored in cli using color groups and easier to debug for me

Reply:
why do we need to see that error all the time in console if you are not debugging? seems like you are going to spam the console

Reply:
If we are here this error does not somehow tracked above. 
If we see it it's a real error need to be fixed ASAP. 
The debug library doesn’t natively support log levels like “info”, “warn”, or “error”. It’s designed to provide debugging output and uses a simple on/off mechanism based on namespaces.
Having it's coloring and ability to use search in console there is no way to leave here debug or console only message.

console and debug are for different purposes. console should work always, debug should work only if namespace is enabled, and allows to filter output by that namespace.

Reply:
got it, I am wondering if we could have used one for both, because this seems like going to be everywhere and its not very nice, even though I understand the reasoning.

As an idea debug with error could be always logged by using error namespace